rotkehlchen/chain/aggregator.py

```diff
@@ -1286,72 +1289,164 @@ def get_evm_manager(
+        the account is active there and if it is, starts tracking it. Returns a list of tuples
+        (chain, account) for each chain where the account was added in.
+        """
+        active_chains = self.check_single_address_activity(
+            address=account,
+            chains=chains,
+        )
+        new_tracked_chains = self.track_evm_address(account, active_chains)
 
-        return filtered_accounts
+        return [(chain, account) for chain in new_tracked_chains]
 
     def add_accounts_to_all_evm(
             self,
             accounts: list[ChecksumEvmAddress],
     ) -> list[tuple[SUPPORTED_EVM_CHAINS, ChecksumEvmAddress]]:
-        """Adds each account for all evm chains depending on our filter.
+        """Adds each account for all evm chain if it is not a contract in ethereum mainnet.
 
         Returns a list of tuples of the address and the chain it was added in.
         """
-        filtered_accounts = self.filter_active_evm_addresses(accounts)
-        added_accounts = []
-        for chain, account in filtered_accounts:
-            try:
-                self.modify_blockchain_accounts(
-                    blockchain=chain,
-                    accounts=[account],
-                    append_or_remove='append',
+        added_accounts: list[tuple[SUPPORTED_EVM_CHAINS, ChecksumEvmAddress]] = []
+        # Distinguish between contracts and EOAs
+        for account in accounts:
+            if self.is_contract(account, SupportedBlockchain.ETHEREUM):
+                added_chains = self.track_evm_address(account, [SupportedBlockchain.ETHEREUM])
+                if len(added_chains) == 1:  # Is always either 1 or 0 since is only for ethereum
+                    added_accounts.append((SupportedBlockchain.ETHEREUM, account))
+            else:
+                chains_to_check = []
+                for chain in typing.get_args(SUPPORTED_EVM_CHAINS):
+                    if account not in self.accounts.get(chain):
+                        chains_to_check.append(chain)
+
+                added_accounts += self.check_chains_and_add_accounts(
+                    account=account,
+                    chains=chains_to_check,
                 )
-            except InputError:
-                log.debug(f'Not adding {account} to {chain} since it already exists')
+
+        return added_accounts
+
+    def detect_evm_accounts(
+            self,
+            progress_handler: Optional['ProgressUpdater'] = None,
+    ) -> list[tuple[SUPPORTED_EVM_CHAINS, ChecksumEvmAddress]]:
+        """
+        Detects user's EVM accounts on different chains and adds them to the tracked accounts.
+        1. Iterates through already added addresses
+        2. Checks whether the address is the chain that it is currently added to
+        3. If it is not, checks whether the address is active in a chain
+        4. If it is active, adds it to the tracked accounts on that chain
```

Comment:
From bullet point 2 and on it does not make sense.

> Checks whether the address is the chain that it is currently added to 
???

 I am rewriting how I understand it.

```suggestion
        2. For each address, assuming it's not a contract, checks which chains it's already in
        3. Get the rest of the chains, and check activity. If active in any of them it tracks the address for that chain.
```

Reply:
Thanks